design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,777 @@
+# End-to-end Testing in Crossplane
+
+* Owner: Predrag Knežević (@pedjak), Lovro Sviben (@lsviben)
+* Reviewers: @negz, TBD
+* Status: Draft
+
+## Background
+
+As a step in maturing Crossplane, we should be thinking about setting up
+processes and methodology that would enable automated end-to-end testing,
+covering all user-facing functionalities.
+
+Currently, as reported in [this issue], there seems to be a lack of e2e tests
+for Crossplane, impacting directly confidence in changes submitted in PRs and
+the quality of published releases.
+
+The following repos contains some sort of e2e tests:
+
+* [crossplane/test], last test change was 9 month ago
+* Crossplane repo itself [crossplane/crossplane e2e], last update was 10 months
+  ago
+* [crossplane/conformance], last change in June, 2021
+
+Apart for the stale development, the existing tests:
+
+* Cover only tiny fraction of all user-facing features
+* Do not cover any error cases
+* Are reported to be flaky sometimes
+
+The tests are written in pure Go, requesting to keep the number of library
+dependencies at minimum.
+
+Contrary to unit tests, end-to-end testing requires very often complex test 
+fixtures (e.g. installing Crossplane, providers, reaching some state in the
+cluster) before we can focus on validating a particular system behavior.
+Defining the fixture only with the help of Go standard library can be
+time-consuming, requiring to write a significant amount of boiler-plate code,
+hence reducing test readability and maintenance, and eventually not very
+motivating for creating new tests. In order to improve the readability, one
+would start looking for patterns and either eventually end up with a custom
+framework or adopt an existing one.
+
+**_NOTE_**: The document uses end-to-end testing term interchangeably
+with [system testing].
+
+## Goals
```

Comment:
I think what we have here are the list of properties we want from an E2E testing framework. Would it also be helpful to agree here on why we want an E2E framework? What isn't working well today? Presumably our overall goal (the reason we're adding E2E tests) is to solve those problems.

Reply:
The proposal is about introducing end-to-end tests that should match the listed goals (or properties). The goals are orthogonal to evaluated frameworks - they are matching more or less with the set goals.

The motivation for the proposal is stated in Background section. Feel free to propose some rephrasing.

Reply:
Let me try to frame what I think is the problem we need to solve. Apologies if some of this is stating the obvious.

Our problem is that our unit tests are not sufficient. We rely on an honor system of folks running manual black-box tests to make sure code works in real-world conditions. These tests:
  * Are not very repeatable, or portable for others to run.
  * Are often only run on one environment - i.e. `kind` on the author's laptop.
  * Are not re-run on future changes to catch regressions.
  * Are likely to miss things like performance degradation.
  * Are usually skipped entirely for innocuous seeming things like Renovate dependency bumps.
  * Often don't cover upgrade scenarios.

So I believe our goal is to have black-box tests that solve the above problems. i.e. Are repeatable, portable, run frequently, and catch performance and upgrade issues. We need to both _introduce_ these tests and _ensure folks keep writing them_.

Where I think we're diverging is that I _do not_ believe solving the above problems requires introducing a new testing framework. That is, I don't think the way we write, structure, and run tests today needs to change. I feel the (stdlib) `testing` and `cmp`  packages that we already use are just as well suited to integration and E2E tests as they are to unit tests. You can do a lot with `t.Run()`, `t.Cleanup()`, etc.

I do think as part of implementing E2E tests we will need some convenience utilities - a library - for doing things like starting a control plane, deploying Crossplane, applying configuration (test inputs), waiting for readiness, etc. Ideally that library would be something that already exists rather than something we roll ourselves.

So what I personally would like to see from this proposal is:

* Align on the problems we're solving - is it what I wrote above? Less? More? Completely different?
* Identify a library that will make _black-box testing Crossplane_ (not writing tests in general) more convenient.
* Identify what tests add the most value - what should we start with?
* Figure out how we're going to ensure folks actually write tests going forward.